### ISSUE TITLE
Add better support for Obj-C protocols [phab:D1079]

### ISSUE THREAD

**Issue by kythe** [google/kythe#1424]
_Created by **salguarnieri** at 2016-09-23 21:41:05_: 

* Add extends edge from class to protocol
* Add overrides edge from method to method decl in protocol
* Add tests for these new edges
* Add test to show that different namespaces for classes and protocols are
  supported.

Also:
Add tests for parameter types in Obj-C

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-09-23 22:59:11_:

Is this the right RangeInCurrentContext to use?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-09-23 22:59:11_:

It's better not to test against signatures without explicit forms specified by the schema. I expect this is probably a tnominal named FwdClass#n.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-09-23 22:59:11_:

Will this ever make sense?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-09-23 22:59:11_:

Will this assertion work now?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-09-23 22:59:11_:

Maybe check that `Super` refers to something here.

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2016-09-23 23:32:53_:

Your suspicion is correct, but I'm not entirely sure I know how to fix the problem.

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2016-09-23 23:32:53_:

no, removed.

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2016-09-23 23:32:53_:

no, removed.

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2016-09-23 23:32:53_:

Done.

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2016-09-23 23:32:53_:

:-\ It shouldn't be, but the tests pass both ways and that doesn't make sense to me.

if (auto SRCC = ExplicitRangeInCurrentContext(SuperRange)) ... all tests pass
if (auto SRCC = RangeInCurrentContext(SCID, SuperRange)) ... all tests pass

And i protocol_impl.m we test

```
...
//- Obj ref ObjImpl
//- Shape ref ShapeProto
//- Bounce ref BounceProto
interface Box : Obj<Shape,  Bounce>
```

It doesn't make sense that we can use RICC for the superclass but not for the protocols.

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-09-23 23:33:44_:

  - comments

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-09-24 00:08:51_:

If this block is entirely removed (the block starting with `if (auto SC=`), the test still passes; this indicates that a different part of the visitor is calling to `recordDeclUseLocation`. Using `RangeInCurrentContext` here is probably emitting an incorrect implicit anchor.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2016-09-24 00:08:51_:

This change looks good.

**Comment by kythe** [comment]
_**salguarnieri** wrote at 2016-09-26 18:39:36_:

  - Explain why we link to superclass impl.

**Comment by kythe** [comment]
_**salguarnieri** commented inline (context lost during migration) at 2016-09-26 18:44:29_:

Yes. Discussed offline and added a comment to the code and a condition to the test to make our decision explicit.

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2016-09-26 19:56:47_

**Comment by kythe** [comment]
_**salguarnieri** closed via commit 197d4c27ec6f242b2e0252a829756703ba39acda at 2016-09-26 20:07:09_
